### PR TITLE
Implement timeformat option for CSV reader

### PR DESCRIPTION
**Pull request by nils-dohop**
This implements a timeformat option for the CSV reader, which comes in useful with weirdly formatted time columns.

Includes a trivial unit test.

### CODE REVIEW

File: src/function/scalar/strftime_format.cpp [hunk 1]
```diff
@@ -1360,6 +1360,13 @@ bool StrpTimeFormat::ParseResult::TryToDate(date_t &result) {
 	return Date::TryFromDate(data[0], data[1], data[2], result);
 }
 
+bool StrpTimeFormat::ParseResult::TryToTime(dtime_t &result) {
```
Comment (hawkfish) [inline src/function/scalar/strftime_format.cpp:1363]: Unfortunately we are missing `bool Time::TryFromTime(hh, mm, ss, us, off, result)`, so this will not fail if the time is invalid. I think we need to move this duplicated code into `time.cpp` and add a call to `IsValidTime` at the end. Among other things I'm seeing that the parsing code for timestamps does not accept `24:00:00`, which is valid. This may be more than you bargained for, so unless you want to wade in deeper, this is fine for now.

File: test/sql/copy/csv/test_date.test [hunk 2]
```diff
@@ -37,4 +37,10 @@ query TT
...
 123	2023-01-02 12:33:00
-124	2023-12-02 11:57:00
\ No newline at end of file
+124	2023-12-02 11:57:00
+
+query TT
+Select * from read_csv('data/csv/custom_date.csv', header=true, timeformat='%m/%d/%Y, %-I:%-M %p', types = ['BIGINT', 'TIME'] );
+----
+123	12:33:00
+124	11:57:00
```
Comment (hawkfish) [inline test/sql/copy/csv/test_date.test:46]: I wish this test had some invalid data too, but the existing test does not so it is fine for now.

**Review by hawkfish — APPROVED** [review]

File: src/function/scalar/strftime_format.cpp (hunk 1 above)
Comment (hawkfish) [reply src/function/scalar/strftime_format.cpp:1363]: The problem with piggybacking off the timestamp parser is that if there is an offset specified in the timestamp format, then the time can be invalid. That could produce unfortunate and possibly unpleasant results.

**Review by hawkfish — COMMENTED** [review]
We need to check the parsed UTC offset is zero. Adding a test for this would be a good idea.

**Review by hawkfish — CHANGES_REQUESTED** [review]
Please change TryToTime to check for zero offset (data[7])

File: src/function/scalar/strftime_format.cpp (hunk 1 above)
Comment (nils-dohop) [reply src/function/scalar/strftime_format.cpp:1363]: I'll take a look.

File: test/sql/copy/csv/test_date.test (hunk 2 above)
Comment (nils-dohop) [reply test/sql/copy/csv/test_date.test:46]: Any tips on how to run just a subset of the unit tests? I'm afraid the docs aren't that forthcoming.

**Comment by nils-dohop** [comment]
Will try to fix suggestions. May need some help though. :-)

File: src/function/scalar/strftime_format.cpp (hunk 1 above)
Comment (nils-dohop) [reply src/function/scalar/strftime_format.cpp:1363]: Fixed all issues here except `24:00:00`. That will need some more consideration.

File: test/sql/copy/csv/test_date.test (hunk 2 above)
Comment (nils-dohop) [reply test/sql/copy/csv/test_date.test:46]: Added a separate file for the failing tests.

**Comment by nils-dohop** [comment]
Python build appears to be failing due to docker issues from what I can see.

**Comment by hawkfish** [comment]
I have a [PR](https://github.com/duckdb/duckdb/pull/11519) to clean up `strftime_format` that might make your life easier.

**Comment by nils-dohop** [comment]
Ah, I think I implemented more or less the same thing here. Have a look.

File: src/common/types/time.cpp [hunk 3]
```diff
@@ -273,6 +273,14 @@ dtime_t Time::FromTime(int32_t hour, int32_t minute, int32_t second, int32_t mic
 	return dtime_t(result);
 }
 
+bool Time::TryFromTime(int32_t hour, int32_t minute, int32_t second, int32_t microseconds, dtime_t &result) {
+	if (!Time::IsValidTime(hour, minute, second, microseconds)) {
```
Comment (hawkfish) [inline src/common/types/time.cpp:277]: The parser guarantees this to be true. There is an asymmetry with dates because dates need to be checked for month length, leap years and so on. I'd merge my PR (it's only two files).

File: src/function/scalar/strftime_format.cpp (hunk 1 above)
Comment (hawkfish) [reply src/function/scalar/strftime_format.cpp:1363]: That turns out to not be a thing. Postgres will cast `24:00:00` from ISO-8601 but not parse it in `to_timestamp` so don't worry about it.

File: src/include/duckdb/common/types/time.hpp [hunk 4]
```diff
@@ -37,6 +37,8 @@ class Time {
 	DUCKDB_API static string ToUTCOffset(int hour_offset, int minute_offset);
 
 	DUCKDB_API static dtime_t FromTime(int32_t hour, int32_t minute, int32_t second, int32_t microseconds = 0);
+	DUCKDB_API static bool TryFromTime(int32_t hour, int32_t minute, int32_t second, int32_t microseconds,
```
Comment (hawkfish) [inline src/include/duckdb/common/types/time.hpp:40]: I don't think this is needed if you use my cleanup.

File: tools/jdbc/generator.py [hunk 5]
```diff
@@ -17,9 +17,8 @@ def function_hook(fn: CppMethod, config):
             'names': ', '.join(names),
             'params': ', '.join(f'{param["type"]} {name}' for param, name in zip(params, names)),
             'return_type': return_type,
-            'short_name': (
-                '_duckdb_jdbc_' + name.replace('Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1', '').replace('1', '')
-            ),
+            'short_name': '_duckdb_jdbc_'
```
Comment (hawkfish) [inline tools/jdbc/generator.py:20]: Is this change intentional?

**Review by hawkfish — CHANGES_REQUESTED** [review]
Getting there! This is feature work so it's probably not going to make the next bug drop on Monday, so why don't you merge in my cleanup PR (which is also tagged as feature work) so we don't collide?

File: src/include/duckdb/common/types/time.hpp (hunk 4 above)
Comment (nils-dohop) [reply src/include/duckdb/common/types/time.hpp:40]: I think it's clean though in the sense that the function ends up in `time.hpp` which mirrors the structure of date and timestamp.

File: tools/jdbc/generator.py (hunk 5 above)
Comment (nils-dohop) [reply tools/jdbc/generator.py:20]: It wasn't my idea, came in from black.

File: src/include/duckdb/common/types/time.hpp (hunk 4 above)
Comment (nils-dohop) [reply src/include/duckdb/common/types/time.hpp:40]: I'll merge your PR though, I guess it effectively increases performance.

Comment (nils-dohop) [reply src/include/duckdb/common/types/time.hpp:40]: I cherry-picked your commit, hope that works for you.

Comment (nils-dohop) [reply src/include/duckdb/common/types/time.hpp:40]: I left these in though, for completeness one could say, time then has the same functions as date and timestamp. I can delete them though if you think that's more suitable.

**Comment by hawkfish** [comment]
OK my PR got merged earlier, so why don't you merge main and push again?

File: src/include/duckdb/common/types/time.hpp (hunk 4 above)
Comment (hawkfish) [reply src/include/duckdb/common/types/time.hpp:40]: Yes please remove - dates and times are not parallel (unfortunately...)

File: src/common/types/time.cpp (hunk 3 above)
Comment (hawkfish) [reply src/common/types/time.cpp:277]: What time ranges are valid depends on the context, so this is not always correct (it is not correct when using a format string for example). So let's just remove this.

**Review by hawkfish — CHANGES_REQUESTED** [review]
Just remove the changes to time.* and merge from main and I think we are good to go.

**Comment by nils-dohop** [comment]
@hawkfish thanks for all the feedback. Should be good to go now.

**Review by hawkfish — APPROVED** [review]
Thanks - LGTM!

**Comment by Mytherin** [comment]
@pdet can you also have a look?

File: test/sql/copy/csv/test_csv_time_format_errors.test [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# name: test/sql/copy/csv/test_csv_time_format_errors.test
+# description: Test reading csv files containing time format errors
+# group: [csv]
+
+statement error
```
Comment (pdet) [inline test/sql/copy/csv/test_csv_time_format_errors.test:5]: Could you add 
```
statement ok
PRAGMA enable_verification
```

to the top of all sql tests and make sure the new option is properly serialized?

File: test/sql/copy/csv/test_csv_time_format_errors.test [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+Column Date_Time is being converted as type TIME
+
+
+statement error
+Select * from read_csv('data/csv/time_format_errors_out_of_bounds.csv', header=true, timeformat='%H:%M', types = ['BIGINT', 'TIME'] );
+----
+Conversion Error: CSV Error on Line: 3
+Error when converting column "Date_Time".
+Could not parse string "23:67" according to format specifier "%H:%M"
+23:67
```
Comment (pdet) [inline test/sql/copy/csv/test_csv_time_format_errors.test:23]: Could you also add tests with the `ignore_errors` option set to true?

Comment (pdet) [reply test/sql/copy/csv/test_csv_time_format_errors.test:23]: Could we also add tests with CSV files that are over one vector size (i.e., > 2048 elements).

**Review by pdet — CHANGES_REQUESTED** [review]
Hi @nils-dohop many thanks for the PR. I think it's looking great.

I just asked for a few tests. I do expect the serializiation test to not work.

For that, you basically will need to add the new option to `src/include/duckdb/storage/serialization/nodes.json` and run `make generate-files`

I also think it would be interesting to add the time mask to the sniffer with some common mask options, but I believe that can also be added later (What do you think @Mytherin ?)

**Comment by pdet** [comment]
Also @Mytherin should this be re-targeted to feature?

**Comment by nils-dohop** [comment]
I'm abandoning this PR. Sorry, but things seem to have changed too much for this to have any future. I'll just work around this instead.

**Comment by nils-dohop** [comment]
I've fixed the remaining comments, @pdet , but based on the other work that was done in the same files this PR is no longer current, if you feel like figuring out what's needed to make it work again, please feel free to merge the commits.

**Comment by Tishj** [comment]
@nils-dohop If you're talking about the python side of things it's much easier than it was before
As explained by this PR <https://github.com/duckdb/duckdb/pull/11202>

The only thing necessary is to edit the `tools/pythonpkg/scripts/connection_methods.json` file, alongside the changes to pyconnection.hpp and pyconnection.cpp

**Comment by nils-dohop** [comment]
@Tishj No, the C++ parts have also moved substantially.
